dev-tools/xtask/src/a4x2_deploy.rs

```diff
@@ -0,0 +1,472 @@
+        retries -= 1;
+        thread::sleep(time::Duration::from_secs(25));
+
+        if retries % 5 == 0 {
+            // I see no reason to error while printing the date
+            let _ = cmd!(sh, "date").run();
+        }
+    }
+
+    if retries == 0 {
+        bail!("timed out waiting for control plane");
+    }
+
+    Ok(())
+}
+
+fn run_tests(sh: &Shell, env: &Environment, args: &RunTestsArgs) -> Result<()> {
+    // TODO: support running end to end tests
+    if args.live_tests {
+        cmd!(sh, "banner 'live tests'").run()?;
+        let _popdir = sh.push_dir(&env.a4x2_dir);
+        let g0ip = get_node_ip(&sh, env, "g0")?;
+
+        let ssh_host = format!("root@{g0ip}");
+        let mut ssh_args_owned = vec!["-i", "../a4x2-ssh-key"];
+        ssh_args_owned.extend_from_slice(&INSECURE_SSH_ARGS);
+        let ssh_args = &ssh_args_owned;
+
+        cmd!(sh, "scp {ssh_args...} live-tests-bundle.tgz {ssh_host}:/zone/oxz_switch/root/root").run()?;
+
+        // If you want any change in functionality for the test runner, update
+        // run-live-tests over in a4x2_package.rs. Don't add it here!
+        let switch_zone_script = r#"
+            set -euxo pipefail
+            tar xvzf live-tests-bundle.tgz
+            cd live-tests-bundle
+            ./run-live-tests
+        "#;
+
+        let remote_script =
+            format!("zlogin oxz_switch bash -c '{switch_zone_script}'");
+
+        // Will error if the live tests fail. This is desired.
+        cmd!(sh, "ssh {ssh_args...} {ssh_host} {remote_script}").run()?;
+    }
+
+    Ok(())
+}
+
+fn teardown_a4x2(sh: &Shell, env: &Environment) -> Result<()> {
+    let _popdir = sh.push_dir(&env.a4x2_dir);
+
+    // destroy a4x2 stuff
+    cmd!(sh, "pfexec ./a4x2 destroy").run()?;
+
+    // destroy the route we added (and any stale ones laying around) Each time
+    // we run `route get` we will get one gateway. If multiple routes have been
+    // added for this IP (by, say, multiple spinup commands without teardowns in
+    // between), then we will need to do this multiple times to fully tear down
+    // all of them. But, I don't like unbounded loops, so we will do at max 10,
+    // which is a number that seems unlikely enough to reach, to me.
+    for _ in 0..10 {
+        let mut route_cmd = cmd!(sh, "pfexec route get 198.51.100.0/24");
+
+        // We get an error code when there is no route (and thus, nothing for
+        // us to delete!)
+        route_cmd.set_ignore_status(true);
+
+        let route = route_cmd.read()?;
+
+        let mut had_gateway = false;
+        for ln in route.lines() {
+            if ln.contains("gateway") {
+                let gateway = ln.split_whitespace().nth(1).ok_or(anyhow!(
+                    "teardown_a4x2: could not get gateway for a4x2 route from line {ln}"
+                ))?;
+
+                cmd!(sh, "pfexec route delete 198.51.100.0/24 {gateway}")
+                    .run()?;
```

Comment:
Worth printing out the fact that we're doing this.

In general I really like the approach where we first gather all operations into a structure, and then perform them in a separate pass. That enables prompting for confirmation, and things like `--dry-run`. Not sure if you want to put the effort into doing that, but if you do I'd really appreciate it :)